lib/trie/trie.go

```diff
@@ -167,675 +177,895 @@ func (t *Trie) Entries() map[string][]byte {
 	return entries(t.root, nil, make(map[string][]byte))
 }
 
-func entries(current Node, prefix []byte, kv map[string][]byte) map[string][]byte {
-	switch c := current.(type) {
-	case *node.Branch:
-		if c.Value != nil {
-			kv[string(codec.NibblesToKeyLE(append(prefix, c.Key...)))] = c.Value
-		}
-		for i, child := range c.Children {
-			entries(child, append(prefix, append(c.Key, byte(i))...), kv)
-		}
-	case *node.Leaf:
-		kv[string(codec.NibblesToKeyLE(append(prefix, c.Key...)))] = c.Value
+func entries(parent Node, prefix []byte, kv map[string][]byte) map[string][]byte {
+	if parent == nil {
+		return kv
+	}
+
+	if parent.Type() == node.LeafType {
+		parentKey := parent.GetKey()
+		fullKeyNibbles := make([]byte, 0, len(prefix)+len(parentKey))
+		fullKeyNibbles = append(fullKeyNibbles, prefix...)
+		fullKeyNibbles = append(fullKeyNibbles, parentKey...)
+		keyLE := string(codec.NibblesToKeyLE(fullKeyNibbles))
+		kv[keyLE] = parent.GetValue()
 		return kv
 	}
 
+	// Branch with/without value
+	branch := parent.(*node.Branch)
+
+	if branch.Value != nil {
+		fullKeyNibbles := make([]byte, 0, len(prefix)+len(branch.Key))
+		fullKeyNibbles = append(fullKeyNibbles, prefix...)
+		fullKeyNibbles = append(fullKeyNibbles, branch.Key...)
+		keyLE := string(codec.NibblesToKeyLE(fullKeyNibbles))
+		kv[keyLE] = branch.Value
+	}
+
+	for i, child := range branch.Children {
+		childPrefix := make([]byte, 0, len(prefix)+len(branch.Key)+1)
+		childPrefix = append(childPrefix, prefix...)
+		childPrefix = append(childPrefix, branch.Key...)
+		childPrefix = append(childPrefix, byte(i))
+		entries(child, childPrefix, kv)
+	}
+
 	return kv
 }
 
-// NextKey returns the next key in the trie in lexicographic order. It returns nil if there is no next key
-func (t *Trie) NextKey(key []byte) []byte {
-	k := codec.KeyLEToNibbles(key)
+// NextKey returns the next key in the trie in lexicographic order.
+// It returns nil if no next key is found.
+func (t *Trie) NextKey(keyLE []byte) (nextKeyLE []byte) {
+	prefix := []byte(nil)
+	key := codec.KeyLEToNibbles(keyLE)
```

Comment:
It does return the key in nibbles format.

Why I kept it as simply `key` is that all the rest of the code uses `key` for the nibbles key. Most internal/private functions only handle nibbles key as well.

I started renaming all `key` to `keyNibbles` to be explicit about its format (like with `keyLE`), but that kind of looked ugly and added a lot of deltas. So in the end I took the approach to have it as `key` for nibbles and `keyLE` for Little Endian formatted keys. What do you think?

As in, I could also rename `key` to `keyNibbles` only at the 'entrypoints' where we deal with both nibbles and little endian keys.

Reply:
Also key fields in nodes (branch, leaf) are just named `Key` although they are in nibbles, so I think it would be good to just keep it as 'key' for nibbles-formatted keys